### PR TITLE
Fix PageView throws Null check error

### LINKED ISSUE #101007: PageView applyViewportDimension throws Null check operator used on a null value
Found this very rare Crashlytics error in production on a Galaxy Note10+. I'm not sure of the trigger or a reproduction.

Looks like it's due to this: https://github.com/flutter/flutter/blob/f07f6fef9ae9156c4531b62ba4a63185528f2884/packages/flutter/lib/src/widgets/page_view.dart#L455

```
Non-fatal Exception: io.flutter.plugins.firebase.crashlytics.FlutterError: Null check operator used on a null value. Error thrown Instance of 'ErrorDescription'.
       at _PagePosition.applyViewportDimension(page_view.dart:453)
       at RenderViewport.performLayout(viewport.dart:1456)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderCustomPaint.performLayout(custom_paint.dart:545)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at ChildLayoutHelper.layoutChild(layout_helper.dart:56)
       at RenderStack._computeSize(stack.dart:570)
       at RenderStack.performLayout(stack.dart:597)
       at RenderObject.layout(object.dart:1887)
       at _RenderLayoutBuilder.performLayout(layout_builder.dart:321)
       at RenderObject.layout(object.dart:1887)
       at MultiChildLayoutDelegate.layoutChild(custom_layout.dart:171)
       at _ScaffoldLayout.performLayout(scaffold.dart:1005)
       at MultiChildLayoutDelegate._callPerformLayout(custom_layout.dart:240)
       at RenderCustomMultiChildLayoutBox.performLayout(custom_layout.dart:403)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at _RenderCustomClip.performLayout(proxy_box.dart:1376)
       at RenderObject.layout(object.dart:1887)
       at ChildLayoutHelper.layoutChild(layout_helper.dart:56)
       at RenderStack._computeSize(stack.dart:570)
       at RenderStack.performLayout(stack.dart:597)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at ChildLayoutHelper.layoutChild(layout_helper.dart:56)
       at RenderStack._computeSize(stack.dart:570)
       at RenderStack.performLayout(stack.dart:597)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderOffstage.performLayout(proxy_box.dart:3430)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at _RenderTheatre.performLayout(overlay.dart:751)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at ChildLayoutHelper.layoutChild(layout_helper.dart:56)
       at RenderStack._computeSize(stack.dart:570)
       at RenderStack.performLayout(stack.dart:597)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderProxyBoxMixin.performLayout(proxy_box.dart:116)
       at RenderObject.layout(object.dart:1887)
       at RenderView.performLayout(view.dart:165)
       at RenderObject._layoutWithoutResize(object.dart:1731)
       at PipelineOwner.flushLayout(object.dart:887)
       at RendererBinding.drawFrame(binding.dart:497)
       at WidgetsBinding.drawFrame(binding.dart:883)
       at RendererBinding._handlePersistentFrameCallback(binding.dart:363)
       at SchedulerBinding._invokeFrameCallback(binding.dart:1144)
       at SchedulerBinding.handleDrawFrame(binding.dart:1081)
       at SchedulerBinding._handleDrawFrame(binding.dart:995)
```

<details>
  <summary>Logs</summary>

```
```

```

[✓] Flutter (Channel stable, 2.10.1, on macOS 12.3 21E230 darwin-arm, locale en-US)
    • Flutter version 2.10.1 at /Users/james/Development/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision db747aa133 (7 weeks ago), 2022-02-09 13:57:35 -0600
    • Engine revision ab46186b24
    • Dart version 2.16.1
    • DevTools version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/james/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)

[✓] VS Code (version 1.65.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.32.0

[✓] Connected device (2 available)
    • Apollo (mobile) • da060dc78ac99566c39cb688206b34cb424a5ee1 • ios            • iOS 15.3.1 19D52
    • Chrome (web)    • chrome                                   • web-javascript • Google Chrome 99.0.4844.84
    ! Error: Apollo is busy: Preparing the watch for development via Apollo. Xcode will continue when Apollo is finished. (code -10)

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!



```

</details>

### PR DESCRIPTION
**Pull request by bleroux**
## Description

This PR fixes a null check error in `PageView`.

The null check was added by the following PR : https://github.com/flutter/flutter/pull/65015

Flutter 3.0 is the first stable release where the null check error surfaces.
It surfaces due to the following PR : https://github.com/flutter/flutter/pull/97971
which added a call to `MediaQuery` in `SrollableState.didChangeDependencies` ([line 462](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/scrollable.dart#L460)) :

```dart
  @override
  void didChangeDependencies() {
    _mediaQueryData = MediaQuery.maybeOf(context);
    _updatePosition();
    super.didChangeDependencies();
  }
```
During startup, especially in release mode, `MediaQuery.of(context).size` might return (0,0). The size will be updated when the native platform reports the actual resolution (several milliseconds later). This seems to be by design for performance reasons (see https://github.com/flutter/flutter/issues/25827#issuecomment-571804641).

For `PageView`, it means that the Viewport dimensions are updated during startup which was not expected with the previous code.

## Related Issue

Fixes https://github.com/flutter/flutter/issues/101007

## Tests

Adds 1 test.

### CODE REVIEW

**Review by Piinks — APPROVED** [review]
LGTM! Thank you for fixing this! @xu-baolin WDYT?

**Comment by xu-baolin** [comment]
Hey， this issue is similar to this #103328, 
We should override `absorb` when the `_PagePosition` was replaced.
```dart
  @override
  void absorb(ScrollPosition other) {
    super.absorb(other);
    assert(_cachedPage == null);

    if (other is! _PagePosition) {
      return;
    }

    if (other._cachedPage != null) {
      _cachedPage = other._cachedPage;
      other._cachedPage = null;
    }
  }
```

**Review by xu-baolin — CHANGES_REQUESTED** [review]
See the comment above.

**Review by xu-baolin — APPROVED** [review]
LGTM，need a second review from @Piinks

**Review by Piinks — APPROVED** [review]
Ah nice @xu-baolin! Thanks @bleroux, LGTM!

**Comment by armandojimenez** [comment]
Anyone knows when is this landing on stable branch>

**Comment by ffamar** [comment]
The issue still exist in the flutter version 3.0.5.
Anyone knows when the patch will be released?

**Comment by alexaung** [comment]
Yes still happening in version 3.0.5

**Comment by tonyshkurenko** [comment]
Is this expected to be fixed in 3.0.5 and it's an issue again or it wasn't included in 3.0.5?

**Comment by bleroux** [comment]
@tonyshkurenko
It was not included in 3.0.5.
For the moment, this fix is available in the master channel and in the 3.3 beta channel.

**Comment by ffamar** [comment]
I confirm, issue fixed in 3.3.0 release.

**Pull request merged**
